src/twisted/test/test_defer.py

```diff
@@ -2191,77 +2297,86 @@ def testLock(self):
         firstUnique = object()
         secondUnique = object()
 
-        controlDeferred = defer.Deferred()
+        controlDeferred: Deferred[object] = Deferred()
 
-        def helper(self, b):
-            self.b = b
-            return controlDeferred
+        result: Optional[object] = None
 
-        resultDeferred = lock.run(helper, self=self, b=firstUnique)
+        def helper(resultValue: object, returnValue: object = None) -> object:
+            nonlocal result
+            result = resultValue
+            return returnValue
+
+        resultDeferred = lock.run(
+            helper, resultValue=firstUnique, returnValue=controlDeferred
+        )
```

Comment:
Use a variable (`nonlocal` in `helper()`) instead of setting an attribute on `self`.